src/main/scala/com/lifeway/aws/lambda/CustomResourceProvider.scala

```diff
@@ -0,0 +1,281 @@
+  }
+
+  def decodeRequest[T](typeDecoder: Decoder[T]): Decoder[Request] = (c: HCursor) => {
+
+    implicit val propDecoder: Decoder[T] = Decoder.decodeString
+      .withErrorMessage("The request resource properties must be a stringified JSON object. Parsing failed.")
+      .emap[T](
+        jsonString =>
+          parser
+            .decode[T](jsonString)(typeDecoder)
+            .fold(
+              error => Left(error.getMessage),
+              obj => Right(obj)
+          )
+      )
+
+    val res = for {
+
+      requestType           <- c.get[String]("RequestType")
+      requestID             <- c.get[String]("RequestId")
+      responseUrl           <- c.get[String]("ResponseURL")
+      resourceType          <- c.get[String]("ResourceType")
+      logicalResourceID     <- c.get[String]("LogicalResourceId")
+      physicalResourceID    <- c.get[Option[String]]("PhysicalResourceId")
+      stackID               <- c.get[String]("StackId")
+      resourceProperties    <- c.get[Option[T]]("ResourceProperties")
+      oldResourceProperties <- c.get[Option[T]]("OldResourceProperties")
+
+    } yield {
+
+      requestType match {
+
+        case RequestTypes.Create =>
+          Right(
+            CreateRequest(
+              requestID,
+              responseUrl,
+              resourceType,
+              logicalResourceID,
+              stackID,
+              resourceProperties
+            )
+          )
+
+        case RequestTypes.Update if physicalResourceID.isDefined =>
+          Right(
+            UpdateRequest(
+              requestID,
+              responseUrl,
+              resourceType,
+              logicalResourceID,
+              stackID,
+              physicalResourceID.get,
+              resourceProperties,
+              oldResourceProperties
+            )
+          )
+
+        case RequestTypes.Update =>
+          Left(
+            DecodingFailure(
+              "The PhysicalResourceId field is required for the Update RequestType",
+              List(DownField("PhysicalResourceId"))
+            )
+          )
+
+        case RequestTypes.Delete if physicalResourceID.isDefined =>
+          Right(
+            DeleteRequest(
+              requestID,
+              responseUrl,
+              resourceType,
+              logicalResourceID,
+              stackID,
+              physicalResourceID.get,
+              resourceProperties
+            )
+          )
+
+        case RequestTypes.Delete =>
+          Left(
+            DecodingFailure(
+              "The PhysicalResourceId field is required for the Delete RequestType",
+              List(DownField("PhysicalResourceId"))
+            )
+          )
+
+        case other =>
+          Left(
+            DecodingFailure(
+              s"Invalid request type ($other). Parsing failed.",
+              List(DownField("RequestType"))
+            )
+          )
+      }
+    }
+
+    // Flatten out the Either
```

Comment:
I don't particularly like this approach, so if you have some Scala magic to make this prettier, let me know.

Reply:
I reworked the encoder to something I like better